File: threading/archer/Advanced/C/Coherency/coherency.c
#include <stdio.h>
#include <math.h>
#include <omp.h>


#define N 64
#define M 1000
#define REPS 10000

void sumsubfs(int); 
double second(void); 

float x[M][N];


int main(int argv, char **argc){

  int i,j,c;
  double time1, time2; 


  for (i=0;i<M;i++) {
    for (j=0;j<N;j++) {
      x[i][j] = 3.142; 
    }
  }


  printf("Offset    Mflop/s\n");

  for (c=0;c<N;c++){
    time1=second();
    for (i=0;i<REPS;i++){
      sumsubfs(c); 
    }
    if (x[0][0] < 0.0) printf("Help!\n"); 
    time2=second();

    printf("%d        %f\n",c-20,((float)(M*REPS))/(1000000.*(time2-time1)));
  }
	   
}


void sumsubfs(int c){

      int i,id;

#pragma omp parallel private(i,id)
      {
        id = omp_get_thread_num(); 
	if (id==0) {
	  
	  for (i=0;i<M;i++){
	    x[i][20]+=3.142;
	  }
	}
	else {
	  if (id==1) {
	    
	    for (i=0;i<M;i++){
	      x[i][c]+=3.142;
	    }

	  }
	}
      }
  

}

File: threading/Advanced/Fortran90/Coherency/coherency.f
c
	integer i,j,c,n,reps
	parameter (reps=10000,n=64,m=1000)
	real x(n,m)
	real*8 time1,second
        common /junk/ x 
c
c
c Initialise vector data
c
        do i=1,m 
           do j=1,n
              x(j,i) =  3.142
           end do
        end do
c
c sum operation
c
        print *, "Offset    Mflop/s"
c
        do c = 1,n
	time1=second()
           do i=1,reps
              call sumsubfs (x,n,c,m)
           end do 
	time1=second()-time1
C
        print *, c-20, "     ",real(m*reps)/(1000000.*time1)
        end do
c
	end
c
c
c
        subroutine sumsubfs (a,n,offset,m)
c
c .. Scalar arguments 
        integer n,offset,m
c
c .. Array arguments
        real a(n,m)
c
c .. Local scalars
        integer i,id

c .. Functions
        integer omp_get_thread_num
c
!$omp parallel  private(i,id) 
c
        id  = omp_get_thread_num()
c
        if (id .eq. 0) then
c
           do i=1,m
              a(20,i)=a(20,i)+3.142 
           enddo 
c
        elseif (id .eq. 1) then 
c
           do i=1,m
              a(offset,i)=a(offset,i)+3.142
           end do 
c
        end if 
c
!$omp end parallel 
c
        return
        end

